estimagic/differentiation/numdiff_np.py

```diff
@@ -0,0 +1,240 @@
+            f: R^m -> R leads to shape (m, ), usually called Gradient
+            f: R -> R^n leads to shape (n, 1), usually called Jacobian
+            f: R^m -> R^n leads to shape (n, m), usually called Jacobian
+
+    """
+    func_kwargs = {} if func_kwargs is None else func_kwargs
+    partialed_func = functools.partial(func, **func_kwargs)
+    f0 = partialed_func(x) if f0 is None else f0
+
+    x_was_scalar = np.isscalar(x)
+    f_was_scalar = np.isscalar(f0)
+
+    x = np.atleast_1d(x)
+    f0 = np.atleast_1d(f0)
+
+    @nan_if_exception
+    @de_scalarize(x_was_scalar)
+    def internal_func(x):
+        return partialed_func(x)
+
+    steps = generate_steps(
+        x=x,
+        method=method,
+        n_steps=n_steps,
+        target="first_derivative",
+        base_steps=base_steps,
+        scaling_factor=scaling_factor,
+        lower_bounds=lower_bounds,
+        upper_bounds=upper_bounds,
+        step_ratio=step_ratio,
+        min_steps=min_steps,
+    )
+
+    evaluation_points = []
+    for step_arr in steps:
+        for i, j in product(range(n_steps), range(len(x))):
+            if np.isfinite(step_arr[i, j]):
+                point = x.copy()
+                point[j] += step_arr[i, j]
+                evaluation_points.append(point)
+            else:
+                evaluation_points.append(np.nan)
+
+    raw_evals = _nan_skipping_batch_evaluator(internal_func, evaluation_points, n_cores)
+
+    evals = np.array(raw_evals).reshape(2, n_steps, len(x), -1)
+    evals = np.transpose(evals, axes=(0, 1, 3, 2))
+    evals = namedtuple_from_kwargs(pos=evals[0], neg=evals[1])
+
+    jac_candidates = {}
+    for m in ["forward", "backward", "central"]:
+        jac_candidates[m] = getattr(fd, f"jacobian_{m}")(evals, steps, f0)
+
+    orders = {
+        "central": ["central", "forward", "backward"],
+        "forward": ["forward", "backward"],
+        "backward": ["backward", "forward"],
+    }
+
+    if n_steps == 1:
+        jac = _consolidate_one_step_derivatives(jac_candidates, orders[method])
+    else:
+        raise NotImplementedError("Extrapolation is not yet implemented.")
+
+    res = jac.flatten() if f_was_scalar else jac
+
+    return res
+
+
+def _consolidate_one_step_derivatives(candidates, preference_order):
+    """Replace missing derivative estimates of preferred method with others.
+
+    Args:
+        candidates (dict): Dictionary with derivative estimates from different methods.
+        preference_order (list): Order on (a subset of) the keys in candidates. Earlier
+            entries are preferred.
+
+    Returns:
+        consolidated (np.ndarray): Array of same shape as input derivative estimates.
+
+    """
+    preferred, others = preference_order[0], preference_order[1:]
+    consolidated = candidates[preferred].copy()
+    for other in others:
+        consolidated = np.where(
+            np.isfinite(consolidated), consolidated, candidates[other]
```

Comment:
Just guessing :)
```suggestion
            np.isnan(consolidated), consolidated, candidates[other]
```

Reply:
You would also have to switch the order of `candidates[other]` and `consolidated`, wouldn't you?

Reply:
Right.